### ISSUE TITLE
Feature Request - container roles

### ISSUE THREAD

**Issue by Nicole-Ridgwell-NMMNHS** [ArctosDB/arctos#4084]
**Is your feature request related to a problem? Please describe.**
I have a new work-study, I need him to be able to scan barcodes to check container labels. Currently, it seems the only way for me to do this is either give him manage container (he is way too new for this) OR scan the barcode into catalog search, open one of the specimen records, and look at the part location in the specimen record.

**Describe what you're trying to accomplish** 
I'd like him to have access to find container, but without the ability to click "Edit this Container"

**Describe the solution you'd like**
Add a new role?

**Describe alternatives you've considered**
Adding this ability to the data entry role, BUT I realize that would require widespread consensus.

**Additional context**
If no one else wants this, we can continue our work-around of using catalog search.

**Comment by mkoo** [comment]
I agree with this request!

**Comment by dustymc** [comment]
@Nicole-Ridgwell-NMMNHS what exactly are you needing to see/review?

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
I need a novice user to be able to scan a barcode and see container label, container type, and the container's position in the container hierarchy.

**Comment by dustymc** [comment]
Thanks.

Suggest we change find containers to manage_specimens access and wrap the links in ``if manage_container ...`` as initially requested. I don't think that actually changes access very much, but it would provide a bit more UI to more users. 

Just flip this to next task if nobody objects in the next week or so.

Alternatives include SQL (but recursive SQL is not pretty and may not work well in the UI, and flatter approaches are - well, flatter) or some new report.

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
I didn't think manage_specimens was appropriate for a novice user?

**Comment by dustymc** [comment]
How novice are we talking about?

https://arctos.database.museum/Admin/user_roles.cfm

I can't think of much reason to set a lower limit on viewing, other than "enter data and view containers" seeming like a strange combination.

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
Student has only been working in the collection a few weeks and only interaction with the database so far is to look up barcodes in catalog search.

Context for this is that we're entering part locations and barcoding from the ground up now that we're finally back in the collection. Sticking barcodes to things is an easy job to give to a new student/volunteer, and we'll be doing a lot of that in the next year.

**Comment by dustymc** [comment]
I'm not sure how this ties in with your initial request, but....

>barcoding from the ground up

Associating parts with a machine-identifiable container is probably the most painful operation involving barcodes. More-exposed partIDs might provide a better pathway. Find something, download....


<img width="405" alt="Screen Shot 2021-11-08 at 10 33 10 AM" src="https://user-images.githubusercontent.com/5720791/140798396-ba28a142-16fd-4cdd-aa55-c85470bd3421.png">

... add a 'stuff_this_into_barcode' column, scan into it, and request a new loader which accepts (partID, barcode, [maybe some container metadata, whatever]) and hooks them together.

I don't think I have any useful advice on higher-level (tray-->shelf) operations - I'd always prefer a live app so you can deal with any problems while you're still holding the container, but if you're not comfortable with the user having manage container then nobody else should be either!

Adding help wanted - community guidance would be useful, how have other collections dealt with this sort of situation?

**Comment by campmlc** [comment]
I like this idea? Use partidentifiers as barcode series? This would
eliminate a number of complex steps.

**Comment by campmlc** [comment]
Or do we still use the barcode column? Either way, this seems like progress.

**Comment by campmlc** [comment]
@Nicole-Ridgwell-NMMNHS 
My solution to your issue is to print out an Excel spreadsheet for the student with the part locations as obtained by you from Parts Table Download, and have them find the samples, stick a barcode on the sample, and stick a barcode on the printed Excel, and perhaps scan the barcode into an Excel file if you trust them (that is more scary).

**Comment by dustymc** [comment]
>Use partidentifiers as barcode series? 

No, I think we've established that these are different THINGS, but the end result is almost the same. I'm suggesting using partidentifiers as a "handle" to the part (the part's container actually, but that's details which can be ignored by almost all users). There's no new big-picture functionality, partID just provides an unambiguous way to say "THAT liver" sample in an external file, even if there are 40 of them associated with the record, and using that to put the part into a container (eg bring it into object tracking) should be a relatively trivial loader.

**Comment by campmlc** [comment]
Yes, I support this.

**Comment by dustymc** [comment]
>that is more scary

See https://handbook.arctosdb.org/documentation/container.html - NOT scanning is terrifying from my perspective (and has lead to lots of problems when the wrong thing is typed). I'd always recommend scanning over typing, no exceptions.

**Comment by campmlc** [comment]
I mean, having the novice student scan without training is scary. Scanning
is absolutely essential, but students need training, otherwise they try to
(mis)type the barcode or they scan into the wrong Excel cell or they double
scan two adjacent barcodes.

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
> request a new loader which accepts (partID, barcode

this would be very useful and cut out a lot of steps for me as I'm digitally adding parts to containers! I'll start an issue.

> print out an Excel spreadsheet

Hadn't considered that. It would be better solution than what I'm doing now. At the moment, I am just barcoding cabinets and drawers so the barcodes are preprinted, I just need the student to have something to check they are putting the barcode on the correct container.

**Comment by dustymc** [comment]
>barcoding cabinets and drawers so the barcodes are preprinted,

Are you using smart/meaningful barcodes or arbitrary/dumb? If the latter and it doesn't much matter which container ends up wearing a particular barcode, you could set this up so they're creating those data rather than checking them.

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
> you could set this up so they're creating those data rather than checking them

They're arbitrary/dumb, but at the moment I have the containers pre-created because I was working on importing our inventory while I was teleworking. I might be able to reverse the process once we get past what I've already done.

**Comment by Jegelewicz** [comment]
We especially need to have access control for claiming barcodes and editing claims because it is too easy to change other institutions claims.

**Comment by dustymc** [comment]
Suggest: disallow edit and delete on claimed barcodes, add a valid flag, but not sure who could change that, or if it could be flipped both ways.

Documentation needs clarified; claims help us all play nice together, but only creating barcodes prevents someone else from creating the same barcodes. (Maybe that should be changed, somehow??)

Probably need to revisit uniqueness of barcodes - it's currently global, preprinted barcodes (plates? - @KyndallH ) might make that untenable.

View-only role may need pushed off to https://github.com/ArctosDB/arctos/issues/3694, it's probably Major Surgery. manage_container could (probably) be split into multiple roles, which might allow also relaxing some of the rules - only experienced people would have access to scary stuff.

**Comment by dustymc** [comment]
Discussed in object tracking meeting. Going next task, high priority for separating roles; anything not role-related needs moved to a dedicated Issue. 

After roles are refined, revisit container rules with an eye towards simplification. 

Proposed roles:

* view_container
   *  access the "container tree" and edit forms, save buttons don't work (or aren't there, but that's less trivial/immediate). Safe to give to about anyone.
   * table container access: select
* manage_container
    *  core of the current role, minus the stuff below. Allows shelving, scanning tubes, moving stuff around freezers, etc.
    * table container access: select, update
* admin_container
    * full access; create containers (including positions), delete containers (including access to the scary barcode swapper). Should be *very* restricted role, available to about one careful and trained user per collection.
    * table container access: select, update, insert, delete
    * "barcode series" access (which needs split off of this Issue and refined)

@campmlc @KyndallH @atrox10 @Jegelewicz @Nicole-Ridgwell-NMMNHS what did I mangle? Are the proposed roles a useful starting point? (I was thinking one more, but can't articulate it so maybe 3 is the right number??)

**Comment by campmlc** [comment]
Currently our students at MSB have ability to create containers, eg freezer
boxes, ad hoc, but that is the source of a lot of error. This will require
all containers and positions  be created in advance by an admin. The
cleaner data are worth it, but this might require some culture or workflow
change in some divisions.
 Other than that this looks good.

**Comment by Jegelewicz** [comment]
Agree with @campmlc BUT I do think that perhaps this

"barcode series" access (which needs split off of this Issue and refined)

might need split off in the future? For now I think it's fine though.

**Comment by dustymc** [comment]
>require some culture or workflow change

Yes, on all counts. Untrained users accessing create tools explains a huge percentage of the problems. We've known that for a very long time, it's where the "don't use this form" stuff on the single-record create form comes from. Putting creation on one (or very few) carefully controlled users provides an unavoidable opportunity to add procedures; create everything as recommended in the [handbook](https://handbook.arctosdb.org/documentation/container.html) and most problems never come to exist.

And FWIW I don't think that will ultimately be a burden - spending a small amount of time up front will save a huge amount of time fixing problems. 

>"barcode series" ... might need split

Why? That seems a central part of container administration to me.

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
These roles look good to me,

**Comment by Jegelewicz** [comment]
Maybe a quick first pass is anyone with manage_collection should also have admin_container?

**Comment by dustymc** [comment]
>manage_collection

I don't think that necessarily has any overlap with "very restricted role, available to about one careful and trained user per collection."

**Comment by campmlc** [comment]
I agree with Teresa that admin_container should only be assigned to operators who already have manage collection, which is also a very restricted role (or should be).
I've reviewed all MSB roles in the previous list.

**Comment by dustymc** [comment]
There are 14 active manage_collection users at MSB. If we're not going to treat the admin role with more care than that then this cannot end with better procedures and fewer rules, and I'm not sure what we're hoping to accomplish.

**Comment by campmlc** [comment]
Yes, and I can't be admin_container for all of them . . . this is the role
of the manage collection admin.

**Comment by dustymc** [comment]
> I can't be admin_container for all of them

Well I've obviously misunderstood the situation. This will need fleshed out before we can discuss potential rule changes.

**Comment by campmlc** [comment]
The proposed admin container role is as follows:
admin_container
full access; create containers (including positions), delete containers (including access to the scary barcode swapper). Should be very restricted role, available to about one careful and trained user per collection.
table container access: select, update, insert, delete
"barcode series" access (which needs split off of this Issue and refined)

MSB has 5 divisions actively using object tracking: MSB:Mamm, MSB:Bird, MSB:Para, MSB:Herp, MSB:Fish. While MSB DGR is the primary "user" in the sense of  tracking the large number o tissues for all of these, at least 4 of the above have collection managers who are also using object tracking in their own divisions, and they will need to be able to create and delete containers and positions. I agree that it is a good idea to remove these permissions from students. But collection managers should have these permissions.
